File: py/p9.py
import math
from typing import Tuple

def triple_from_sum(sum: int) -> Tuple[int, int, int]:
    hs = sum // 2
    for m in range(2, math.ceil(math.sqrt(hs))):
        if hs % m == 0:
            for mn in range(m + 1 + (m%2), 2 * m, 2):
                if hs % mn == 0:
                    n = mn - m
                    d = hs // (m * mn)
                    a = (m*m - n*n) * d
                    b = 2 * m * n * d
                    c = (m*m + n*n) * d
                    return a, b, c
    return 0, 0, 0

s = 1000
a, b, c = triple_from_sum(s)
print(a * b * c)

File: rs/p9.rs
use std::cmp;

fn main () {
    let s = 1000;
    let triple = triple_from_sum(s);
    match triple {
        (a, b, c) => println!("{}", a * b * c),
    }
}

fn triple_from_sum(s: u64) -> (u64, u64, u64) {
    let hs = s / 2;

    // exactly one of m, n is even and gcd(m,n) = 1
    for m in 2..(hs as f64).sqrt().ceil() as u64 {
        if hs % m == 0 {
            for mn in ((m + 1 + (m % 2))..(2 * m)).step_by(2) {
                if hs % mn == 0 && gcd(m, mn) == 1 {
                    let n = mn - m;
                    let d = hs / (m * mn);
                    let a = (m * m - n * n) * d;
                    let b = 2 * m * n * d;
                    let c = (m * m + n * n) * d;
                    return (a, b, c);
                }
            }
        }
    }
    (0, 0, 0)
}

fn gcd(x: u64, y: u64) -> u64 {
    if x == 0 && y == 0 { panic!("the numbers cannot both be zero") }
    else if x == 0 { y }
    else if y == 0 { x }
    else {
        let l = cmp::max(x, y);
        let mut s = cmp::min(x, y);
        let mut r = l % s;
        while r > 0 {
            let rem = s % r;
            s = r;
            r = rem;
        }
        s
    }
}
